File: src/update/registry.ts
import type { Logger } from '../core/contracts.js';

// npm registry probe for the latest published version (§7). Reads the package's
// `latest` dist-tag document. Mirrors usageService's HTTP-over-injected-fetch style
// (fetchFn injected, timeout via AbortController, NEVER throws): any failure — offline,
// timeout, non-OK status, malformed JSON — resolves to null so the caller silently skips.

const REGISTRY_URL = 'https://registry.npmjs.org/discord-agent-bridge/latest';
const DEFAULT_TIMEOUT_MS = 5000;

export interface FetchLatestOptions {
  timeoutMs?: number;
  userAgent?: string;
  logger?: Logger;
}

// Resolve the latest published version string, or null on ANY failure. fetchFn is
// injected so tests never touch the network; the real caller passes global fetch.
export async function fetchLatestVersion(
  fetchFn: typeof fetch,
  opts: FetchLatestOptions = {},
): Promise<string | null> {
  const timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetchFn(REGISTRY_URL, {
      signal: controller.signal,
      headers: {
        Accept: 'application/json',
        ...(opts.userAgent ? { 'User-Agent': opts.userAgent } : {}),
      },
    });
    if (!res.ok) {
      opts.logger?.warn('update check: registry returned a non-OK status', { status: res.status });
      return null;
    }
    let body: unknown;
    try {
      body = await res.json();
    } catch (err) {
      opts.logger?.warn('update check: registry returned invalid JSON', { error: String(err) });
      return null;
    }
    const version = (body as { version?: unknown } | null)?.version;
    if (typeof version !== 'string' || version.length === 0) {
      opts.logger?.warn('update check: registry response has no version field');
      return null;
    }
    return version;
  } catch (err) {
    // Offline, DNS failure, or the timeout abort — all silent skips (never-throws).
    opts.logger?.warn('update check: registry request failed', { error: String(err) });
    return null;
  } finally {
    clearTimeout(timer);
  }
}

File: swift/Sources/DiscordAgentBridge/Update/Registry.swift
import Foundation

// npm registry probe for the latest published version (TS `src/update/registry.ts`).
// Injected fetch; NEVER throws — any failure resolves to nil so the caller silently skips.

public let updateRegistryURL = URL(string: "https://registry.npmjs.org/discord-agent-bridge/latest")!
public let updateRegistryDefaultTimeoutMs = 5000

/// Injectable HTTP GET: returns body data + HTTP status (or throws).
public typealias UpdateHTTPGet = @Sendable (URL, [String: String], TimeInterval) async throws -> (Data, Int)

/// Default URLSession-backed GET with timeout.
public let defaultUpdateHTTPGet: UpdateHTTPGet = { url, headers, timeout in
    var req = URLRequest(url: url, timeoutInterval: timeout)
    req.httpMethod = "GET"
    for (k, v) in headers { req.setValue(v, forHTTPHeaderField: k) }
    let (data, resp) = try await URLSession.shared.data(for: req)
    let status = (resp as? HTTPURLResponse)?.statusCode ?? -1
    return (data, status)
}

public struct FetchLatestOptions: Sendable {
    public var timeoutMs: Int
    public var userAgent: String?
    public var registryURL: URL
    public var httpGet: UpdateHTTPGet

    public init(
        timeoutMs: Int = updateRegistryDefaultTimeoutMs,
        userAgent: String? = nil,
        registryURL: URL = updateRegistryURL,
        httpGet: @escaping UpdateHTTPGet = defaultUpdateHTTPGet
    ) {
        self.timeoutMs = timeoutMs
        self.userAgent = userAgent
        self.registryURL = registryURL
        self.httpGet = httpGet
    }
}

/// Resolve the latest published version string, or nil on ANY failure.
public func fetchLatestVersion(opts: FetchLatestOptions = FetchLatestOptions()) async -> String? {
    var headers: [String: String] = ["Accept": "application/json"]
    if let ua = opts.userAgent, !ua.isEmpty {
        headers["User-Agent"] = ua
    }
    let timeout = max(0.001, Double(opts.timeoutMs) / 1000.0)
    let data: Data
    let status: Int
    do {
        (data, status) = try await opts.httpGet(opts.registryURL, headers, timeout)
    } catch {
        return nil
    }
    guard status >= 200, status < 300 else { return nil }
    guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let version = obj["version"] as? String,
          !version.isEmpty
    else { return nil }
    return version
}
